src/ucp/core/ucp_rkey.c

```diff
@@ -213,61 +218,507 @@ UCS_PROFILE_FUNC(ssize_t, ucp_rkey_pack_memh,
+    size_t tl_mkey_size, global_id_size;
+    void *tl_mkey_buf, *global_id_buf;
+    size_t tl_mkey_data_size;
+
+    ucs_log_indent(1);
+
+    ucp_memh_common_pack(memh, &p, UCP_MEMH_BUFFER_FLAG_EXPORTED,
+                         memh_info_size);
+
+    *ucs_serialize_next(&p, uint64_t) = address;
+    *ucs_serialize_next(&p, uint64_t) = length;
+    *ucs_serialize_next(&p, uint64_t) = context->uuid;
+    *ucs_serialize_next(&p, uint64_t) = memh->reg_id;
+
+    memh_info_packed_size = UCS_PTR_BYTE_DIFF(buffer, p);
+    ucs_assertv(memh_info_size == memh_info_packed_size,
+                "memh_info_size %zu memh_info_packed_size %zu",
+                memh_info_size, memh_info_packed_size);
+
+    ucs_for_each_bit(md_index, context->export_md_map[memh->mem_type]) {
+        md_attr           = &tl_mds[md_index].attr;
+        tl_mkey_data_size = ucp_memh_exported_tl_mkey_packed_size(context,
+                                                                  md_index);
+        ucp_memh_info_size_pack(&p, tl_mkey_data_size);
+
+        tl_mkey_size = md_attr->exported_mkey_packed_size;
+        ucs_assertv((tl_mkey_size <= UINT8_MAX) && (tl_mkey_size != 0),
+                    "tl_mkey_size %zu", tl_mkey_size);
+        *ucs_serialize_next(&p, uint8_t) = tl_mkey_size;
+
+        tl_mkey_buf = ucs_serialize_next_raw(&p, void, tl_mkey_size);
+        status      = uct_md_mkey_pack_v2(tl_mds[md_index].md,
+                                          uct_memhs[md_index], &params,
+                                          tl_mkey_buf);
+        if (status != UCS_OK) {
+            result = status;
+            goto out;
+        }
+
+        global_id_size = ucp_memh_global_id_packed_size(md_attr);
+
+        *ucs_serialize_next(&p, uint8_t) = global_id_size;
+
+        global_id_buf = ucs_serialize_next_raw(&p, void, global_id_size);
+        memcpy(global_id_buf, md_attr->global_id, global_id_size);
+
+        ucs_trace("exported mkey[%d]=%s for md[%d]=%s",
+                  ucs_bitmap2idx(context->export_md_map[memh->mem_type],
+                                 md_index),
+                  ucs_str_dump_hex(p, tl_mkey_size, buf, sizeof(buf),
+                                   SIZE_MAX),
+                  md_index, tl_mds[md_index].rsc.md_name);
+    }
+
+    result = UCS_PTR_BYTE_DIFF(buffer, p);
+
+out:
+    ucs_log_indent(-1);
+    return result;
+}
+
+static ucp_md_map_t ucp_find_md_map(ucp_context_h context,
+                                    const void *global_id_buf,
+                                    size_t global_id_size)
+{
+    ucp_md_map_t md_map = 0;
+    ucp_md_index_t md_index;
+    uct_md_attr_v2_t *md_attr;
+    size_t global_id_cmp_size;
+
+    for (md_index = 0; md_index < context->num_mds; md_index++) {
+        md_attr            = &context->tl_mds[md_index].attr;
+        global_id_cmp_size = ucs_min(ucp_memh_global_id_packed_size(md_attr),
+                                     global_id_size);
```

Comment:
can it lead to some weird collisions when some mds will match even so they are not supposed to?
because of different length